stdlib/public/Concurrency/TaskLocal.swift

```diff
@@ -0,0 +1,146 @@
+    return TaskLocalScopeBuilder(keyType: Key.self, value: value)
+  }
+
+  // bind execute
+  public static func withLocal<Key, BodyResult>(
+    _ key: KeyPath<TaskLocalValues, Key>,
+    boundTo value: Key.Value,
+    body: @escaping () async -> BodyResult
+  ) async -> BodyResult where Key: TaskLocalKey { // TODO: cannot express with rethrows
+    let bindings = await self.withLocal(key, boundTo: value)
+
+    let task = await _runChildTaskWithTaskLocal(bindings: bindings) { () async throws -> BodyResult in
+      await body()
+    }
+
+    return await try! Task.Handle(task: task).get() // handle cannot throw, the body never will
+  }
+
+  // bind execute
+  // Throwing version.
+  public static func withLocal<Key, BodyResult>(
+    _ key: KeyPath<TaskLocalValues, Key>,
+    boundTo value: Key.Value,
+    body: @escaping () async throws -> BodyResult
+  ) async throws -> BodyResult where Key: TaskLocalKey { // TODO: cannot express with rethrows
+    let bindings = await self.withLocal(key, boundTo: value)
+
+    let task = await _runChildTaskWithTaskLocal(bindings: bindings) { () async throws -> BodyResult in
+      await try body()
+    }
+
+    return await try Task.Handle(task: task).get()
+  }
+}
+
+public struct TaskLocalScopeBuilder {
+  var bindings: [(AnyTaskLocalKey, Any)]
+
+  init<Key>(keyType: Key.Type, value: Key.Value) where Key: TaskLocalKey {
+    self.bindings = [(AnyTaskLocalKey(keyType), value)]
+  }
+
+  public static func withLocal<Key>(
+    _ key: KeyPath<TaskLocalValues, Key>,
+    boundTo value: Key.Value
+  ) async -> TaskLocalScopeBuilder where Key: TaskLocalKey {
+    fatalError("\(#function) not yet implemented")
+  }
+
+  // bind execute
+  public static func withLocal<Key, BodyResult>(
+    _ key: KeyPath<TaskLocalValues, Key>,
+    boundTo value: Key.Value,
+    body: () async throws -> BodyResult
+  ) async throws -> BodyResult where Key: TaskLocalKey { // TODO: cannot express with rethrows
+    fatalError("\(#function) not yet implemented")
+  }
+}
+
+// ==== ------------------------------------------------------------------------
+
+/// A type-erased `TaskLocalKey` used when iterating through the `Baggage` using its `forEach` method.
+struct AnyTaskLocalKey {
+  /// The key's type represented erased to an `Any.Type`.
+  public let keyType: Any.Type
+
+  init<Key>(_ keyType: Key.Type) where Key: TaskLocalKey {
+    self.keyType = keyType
+  }
+}
+
+extension AnyTaskLocalKey: Hashable {
+  static func ==(lhs: AnyTaskLocalKey, rhs: AnyTaskLocalKey) -> Bool {
+    return ObjectIdentifier(lhs.keyType) == ObjectIdentifier(rhs.keyType)
+  }
+
+  func hash(into hasher: inout Hasher) {
+    hasher.combine(ObjectIdentifier(self.keyType))
+  }
+}
+
+// ==== ------------------------------------------------------------------------
+
+@_silgen_name("swift_task_localSet")
+public func _taskLocalSet/*<Key>*/(
+  _ task: Builtin.NativeObject,
+  keyType: Any.Type/*Key.Type*/,
+  value: Any/*Key.Value*/
```

Comment:
Might want to take this as __owned and then you can init-with-take to store it

Reply:
Hmm... I see, need to read up a bit on those -- thanks! So owned and copy initialize when I store 🤔

Reply:
Okey owned seems great 👍